File: Atcoder_dp/N_Slimes.py
<<<<<<< HEAD
import sys
sys.setrecursionlimit(10000)

I = lambda: input()
II = lambda: int(input())
LII = lambda: list(map(int, input().split()))
=======
I = lambda: input()
II = lambda: int(input())
MII = lambda: map(int, input().split())
LI = lambda: list(input().split())
LII = lambda: list(map(int, input().split()))
LGMII = lambda: map(lambda x: int(x) - 1, input().split())
LGLII = lambda: list(map(lambda x: int(x) - 1, input().split()))

>>>>>>> 2f6612852ebc014714fb29b63627e467478aa78c

inf = float('inf')

n = II()
<<<<<<< HEAD
slimes = LII()

# Compute prefix sums
pref = [0] * (n + 1)
for i in range(n):
    pref[i + 1] = pref[i] + slimes[i]

def get_sum(i, j):
    return pref[j + 1] - pref[i]

dp = [[-1] * n for _ in range(n)]

def rec(i, j):
    if i == j:
        return 0
    if dp[i][j] != -1:
        return dp[i][j]

    min_cost = inf
    for k in range(i, j):
        cost = rec(i, k) + rec(k + 1, j) + get_sum(i, j)
        if cost < min_cost:
            min_cost = cost

    dp[i][j] = min_cost
    return min_cost

print(rec(0, n - 1))
=======
s = LII()

prefix = [0] * (n + 1)
for i in range(n):
    prefix[i + 1] = prefix[i] + s[i]

def get_sum(l, r):
    return prefix[r + 1] - prefix[l]

from functools import lru_cache

@lru_cache(None)
def go(l, r):
    if l == r:
        return 0
    res = inf
    for k in range(l, r):
        res = min(res, go(l, k) + go(k + 1, r) + get_sum(l, r))
    return res

print(go(0, n - 1))
>>>>>>> 2f6612852ebc014714fb29b63627e467478aa78c

File: Atcoder_dp/N_Slimes.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

const long long INF = 1e18;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    cin >> n;
    vector<long long> s(n);
    for (int i = 0; i < n; ++i) cin >> s[i];
    // Prefix sum
    vector<long long> prefix(n + 1, 0);
    for (int i = 0; i < n; ++i) prefix[i + 1] = prefix[i] + s[i];
    // dp[l][r]: min cost to merge slimes from l to r (inclusive)
    vector<vector<long long>> dp(n, vector<long long>(n, 0));
    for (int len = 2; len <= n; ++len) {
        for (int l = 0; l + len - 1 < n; ++l) {
            int r = l + len - 1;
            dp[l][r] = INF;
            for (int k = l; k < r; ++k) {
                dp[l][r] = min(dp[l][r], dp[l][k] + dp[k + 1][r] + prefix[r + 1] - prefix[l]);
            }
        }
    }
    cout << dp[0][n - 1] << '\n';
    return 0;
}
